### PR TITLE
Cudnn2

### PR DESCRIPTION
**Pull request by daemonmaker**
This is to replace my [previous pull request that adds infer_shape to the cudnn opts](https://github.com/Theano/Theano/pull/2312#issuecomment-68911316). Per Arnaud's instruction I created a new branch and cherry picked my changes. All of the appropriate code appears to be there and the tests pass locally. I am keeping the original branch in case I made any mistakes though.

### CODE REVIEW

File: theano/sandbox/cuda/dnn.py [hunk 1]
```diff
@@ -423,6 +453,40 @@ def make_node(self, img, topgrad, desc, h, w):
...
+        w = shape[0][3]  # Width of input feature maps
+        kh = shape[1][2]  # Height of each filter
+        kw = shape[1][3]  # Width of each filter
+        out3 = kh
+        out4 = kw
+
+        desc = node.inputs[2].owner.op
+        sh, sw = desc.subsample
+
+        # We don't have the information necessary, namely the weight size so
```
Comment (abergeron) [inline theano/sandbox/cuda/dnn.py]: That's not actually true: `node.inputs[3]` is weight height and `node.inputs[4]` is weight width

File: theano/sandbox/cuda/dnn.py [hunk 2]
```diff
@@ -477,6 +541,38 @@ def make_node(self, kern, topgrad, desc, h, w):
...
+    def infer_shape(self, node, shape):
+        padh = 0
+        padw = 0
+
+        desc = node.inputs[2].owner.op
+        sh, sw = desc.subsample
+
+        # We don't have the information necessary, namely the image size so
+        # we cannot infer the shape
+        if sh != 1 or sw != 1:
```
Comment (abergeron) [inline theano/sandbox/cuda/dnn.py]: Once again not true anymore.

File: theano/sandbox/cuda/dnn.py [hunk 3]
```diff
@@ -368,6 +368,36 @@ def connection_pattern(self, node):
...
+        nb = shape[1][0]  # Number of output feature maps
+        kh = shape[1][2]  # Height of each filter
+        kw = shape[1][3]  # Width of each filter
+        padh = 0
+        padw = 0
+        if (
+            not node.inputs[2].owner
+            or not isinstance(node.inputs[2].owner.op, GpuDnnConvDesc)
+        ):
+            raise theano.tensor.basic.ShareError("case not implemented and probably not needed")
```
Comment (abergeron) [inline theano/sandbox/cuda/dnn.py]: I fear that this case might get hit very often as the inputs to GpuDnnConvDesc will be mostly constants so it'll get contstant-folded.  To get around this you might need some GpuDnnShape Op or just use ctypes to forage through the C descriptor.

**Comment by abergeron** [comment]
I'll merge this even with the possible problems, since it's still better than nothing.  The change should be done in a subsequent PR.

**Comment by nouiz** [comment]
Just to be sure, the problems are just that we do not always infer the
shape when we could? Or do this cause to infer wrong shape?

**Comment by abergeron** [comment]
It does not always infer the shape.

**Comment by nouiz** [comment]
The constant folding should not be such a problem. If the user put the CPU
operation, it is that version of infer_shape that will be used.

If the user set the dnn operation, then we will do a first pass of infer
shape before constant folding happen. Then we try to keep the simplest/best
infer shape fi there is optimization that change those dnn operation.

**Comment by daemonmaker** [comment]
In that case should I plan to revisit this issue or not?

**Comment by nouiz** [comment]
I do not think this need change. I think this is fine like this.

thanks.

**Comment by abergeron** [comment]
For the gradient cases, you should.  The forward case, probably not.

2015-01-07 15:14 GMT-05:00 daemonmaker <notifications@github.com>:

**Comment by daemonmaker** [comment]
Okay, I will plan to revisit for the gradient case only.

**Pull request merged**